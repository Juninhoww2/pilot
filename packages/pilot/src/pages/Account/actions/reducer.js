import {
  __,
  always,
  anyPass,
  applySpec,
  equals,
  find,
  gte,
  merge,
  not,
  path,
  pathOr,
  pluck,
  pipe,
  prop,
  propEq,
  when,
} from 'ramda'

import {
  ACCOUNT_RECEIVE,
  COMPANY_RECEIVE,
  FEE_PRESET_RECEIVE,
  LOGIN_FAIL,
  LOGIN_RECEIVE,
  LOGIN_REQUEST,
  LOGOUT_RECEIVE,
  RECIPIENT_RECEIVE,
  RECIPIENT_BALANCE_RECEIVE,
  GET_ACQUIRERS_RESPONSE,
} from './actions'

const getBalance = applySpec({
  available: path(['withdrawal', 'maximum']),
  waitingFunds: path(['balance', 'waiting_funds', 'amount']),
})

const initialState = {
  defaultRecipient: {},
  loading: false,
  sessionId: null,
}

export default function loginReducer (state = initialState, action) {
  switch (action.type) {
    case ACCOUNT_RECEIVE: {
      return merge(state, {
        error: null,
        user: action.payload,
      })
    }

    case COMPANY_RECEIVE: {
      return merge(state, {
        company: action.payload,
        error: null,
        loading: false,
      })
    }

    case LOGIN_REQUEST: {
      return merge(state, {
        error: null,
        loading: true,
        user: null,
      })
    }
    case LOGIN_FAIL: {
      return merge(
        state,
        {
          client: null,
          error: action.payload,
          loading: false,
          sessionId: null,
          user: null,
        }
      )
    }
    case LOGIN_RECEIVE: {
      return merge(
        state,
        {
          client: action.payload,
          error: null,
          loading: false,
          sessionId: action.payload.authentication.session_id,
        }
      )
    }

    case LOGOUT_RECEIVE: {
      return initialState
    }

    case RECIPIENT_BALANCE_RECEIVE: {
      return merge(
        state,
        {
          balance: getBalance(action.payload),
        }
      )
    }

    case RECIPIENT_RECEIVE: {
      return merge(
        state,
        {
          defaultRecipient: action.payload,
        }
      )
    }

    case FEE_PRESET_RECEIVE: {
      return merge(
        state,
        {
          defaultRecipient: {
            ...state.defaultRecipient,
            feePreset: action.payload,
          },
        }
      )
    }

    case GET_ACQUIRERS_RESPONSE: {
      return merge(state, {
        acquirers: action.payload,
      })
    }

    default:
      return state
  }
}

const getAntifraudCost = pipe(
  pathOr([], ['gateway', 'live', 'antifraud_cost']),
  find(propEq('name', 'pagarme')),
  prop('cost')
)

const notDefaultInstallments = pipe(
  pluck('installment'),
  anyPass([equals([1, 2, 7]), equals([1])]),
  not
)

const CREDIT_CARD_MDRS_TYPES = [null, 'credit_card']
const findCreditCardMDR = mdrs => mdrs.find(({
  payment_method: paymentMethod,
}) => CREDIT_CARD_MDRS_TYPES.includes(paymentMethod))

const getInstallments = pipe(
  findCreditCardMDR,
  pathOr([], ['installments']),
  when(notDefaultInstallments, always([]))
)

const getFees = pipe(
  prop('pricing'),
  applySpec({
    anticipation: path(['psp', 'live', 'anticipation']),
    antifraud: getAntifraudCost,
    boleto: path(['gateway', 'live', 'boletos', 'payment_fixed_fee']),
    gateway: path(['gateway', 'live', 'transaction_cost', 'credit_card']),
    installments: pipe(
      pathOr([], ['psp', 'live', 'mdrs']),
      getInstallments
    ),
    pix: applySpec({
      paymentFixedFee: path(['gateway', 'live', 'pix', 'payment_fixed_fee']),
      paymentSpreadFee: path(['gateway', 'live', 'pix', 'payment_spread_fee']),
    }),
    transfer: path(['transfers', 'ted']),
  })
)

export const selectCompanyFees = ({ company, defaultRecipient = {} }) => {
  const companyFees = getFees(company)

  if (!defaultRecipient.feePreset) {
    return companyFees
  }

  if (defaultRecipient.anticipation_type === 'compulsory') {
    companyFees.anticipation = defaultRecipient.feePreset.anticipation_rate
  }

  return {
    ...companyFees,
    installments: getInstallments(defaultRecipient.feePreset.mdrs),
  }
}

// The value 1_000_000 is set by the Risk Team on the property
// `company.transfer.blocked_balance_amount` when the company is
// not allowed to do withdraws.
const BLOCKED_AMOUNT_LIMIT = 100000000

export const selectHasBlockedWithdraw = pipe(
  pathOr(0, ['transfers', 'blocked_balance_amount']),
  gte(__, BLOCKED_AMOUNT_LIMIT)
)

export const selectAnticipationType = ({ company, defaultRecipient }) => {
  if (company.capabilities.allow_transaction_anticipation) {
    return 'compulsory'
  }

  return defaultRecipient.automatic_anticipation_type
}
