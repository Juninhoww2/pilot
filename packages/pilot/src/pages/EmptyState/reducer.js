import {
  ADD_ONBOARDING_ANSWERS,
  FETCHING_ONBOARDING_ANSWERS,
  ONBOARDING_ANSWERS_RECEIVE,
  ONBOARDING_ANSWERS_RESET,
  ONBOARDING_ANSWERS_FAIL,
  SKIP_ONBOARDING,
} from './actions'
import isOnboardingComplete from '../../validation/isOnboardingComplete'

const makeInitialState = () => ({
  error: null,
  loading: false,
  onboardingAnswers: undefined,
  skippedOnboarding: false,
})

export default function welcomeReducer (state = makeInitialState(), action) {
  switch (action.type) {
    case FETCHING_ONBOARDING_ANSWERS: {
      return {
        ...state,
        error: null,
        loading: true,
        onboardingAnswers: undefined,
      }
    }
    case ONBOARDING_ANSWERS_RECEIVE: {
      const {
        payload,
      } = action

      return {
        ...state,
        error: null,
        loading: false,
        onboardingAnswers: payload,
      }
    }
    case ADD_ONBOARDING_ANSWERS: {
      const {
        payload,
      } = action

      return {
        ...state,
        onboardingAnswers: {
          ...state.onboardingAnswers,
          ...payload,
        },
      }
    }
    case ONBOARDING_ANSWERS_RESET: {
      return makeInitialState()
    }

    case ONBOARDING_ANSWERS_FAIL:
      return {
        ...state,
        error: action.payload,
        loading: false,
      }

    case SKIP_ONBOARDING:
      return {
        ...state,
        skippedOnboarding: true,
      }

    default: {
      return state
    }
  }
}

export const shouldSkipOnboarding = ({
  error,
  onboardingAnswers,
  skippedOnboarding,
}) => !!error || isOnboardingComplete(onboardingAnswers) || skippedOnboarding
