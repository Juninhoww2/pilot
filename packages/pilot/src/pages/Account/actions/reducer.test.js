import { selectCompanyFees } from './reducer'

/* eslint-disable sort-keys */
const companyFactory = mdrs => ({
  pricing: {
    gateway: {
      test: {
        transaction_cost: {
          credit_card: 50,
          debit_card: 50,
          boleto: 0,
        },
        transaction_spread: {
          credit_card: 0,
          debit_card: 1.5,
          boleto: 0,
        },
        minimum_monthly_payment: 0,
        antifraud_cost: [
          {
            name: 'pagarme',
            cost: 70,
          },
        ],
        boletos: {
          payment_fixed_fee: 380,
          payment_spread_fee: 0,
        },
      },
    },
    psp: {
      test: {
        anticipation: 3.14,
        mdrs,
      },
    },
    transfers: {
      credito_em_conta: 0,
      ted: 367,
      doc: 367,
      inter_recipient: 0,
    },
  },
})
/* eslint-enable sort-keys */

describe('Account Reducer', () => {
  describe('selectCompanyFees', () => {
    describe('when company has 1, 2 and 7 installments', () => {
      const company = companyFactory([{
        installments: [
          { installment: 1, mdr: 3.79 },
          { installment: 2, mdr: 4.19 },
          { installment: 7, mdr: 4.59 },
        ],
        payment_method: 'credit_card',
      }])

      const fees = selectCompanyFees(company)

      it('should return the correct response', () => {
        expect(fees).toMatchSnapshot()
      })
    })

    describe('when company has 1 installments', () => {
      const company = companyFactory([{
        installments: [{ installment: 1, mdr: 3.79 }],
        payment_method: 'credit_card',
      }])

      const fees = selectCompanyFees(company)

      it('should return the correct response', () => {
        expect(fees).toMatchSnapshot()
      })
    })

    describe('when company has 1, 2 and 7 installments and payment_method is null', () => {
      const company = companyFactory([
        {
          installments: [],
          payment_method: 'debit_card',
        }, {
          installments: [
            { installment: 1, mdr: 3.79 },
            { installment: 2, mdr: 4.19 },
            { installment: 7, mdr: 4.59 },
          ],
          payment_method: null,
        }])

      const fees = selectCompanyFees(company)

      it('should return the correct response', () => {
        expect(fees).toMatchSnapshot()
      })
    })

    describe('when company has other installments types', () => {
      const company = companyFactory([{
        installments: [
          { installment: 1, mdr: 3.79 },
          { installment: 4, mdr: 4.19 },
          { installment: 8, mdr: 4.59 },
        ],
        payment_method: 'credit_card',
      }])

      const fees = selectCompanyFees(company)

      it('should return the correct response', () => {
        expect(fees).toMatchSnapshot()
      })
    })
  })
})

