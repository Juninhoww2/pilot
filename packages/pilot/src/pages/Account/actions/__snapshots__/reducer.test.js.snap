// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Account Reducer selectCompanyFees when company has 1 installments should return the correct response 1`] = `
Object {
  "anticipation": 3.14,
  "antifraud": 70,
  "boleto": 380,
  "gateway": 50,
  "installments": Array [
    Object {
      "installment": 1,
      "mdr": 3.79,
    },
  ],
  "transfer": 367,
}
`;

exports[`Account Reducer selectCompanyFees when company has 1, 2 and 7 installments and payment_method is null should return the correct response 1`] = `
Object {
  "anticipation": 3.14,
  "antifraud": 70,
  "boleto": 380,
  "gateway": 50,
  "installments": Array [
    Object {
      "installment": 1,
      "mdr": 3.79,
    },
    Object {
      "installment": 2,
      "mdr": 4.19,
    },
    Object {
      "installment": 7,
      "mdr": 4.59,
    },
  ],
  "transfer": 367,
}
`;

exports[`Account Reducer selectCompanyFees when company has 1, 2 and 7 installments should return the correct response 1`] = `
Object {
  "anticipation": 3.14,
  "antifraud": 70,
  "boleto": 380,
  "gateway": 50,
  "installments": Array [
    Object {
      "installment": 1,
      "mdr": 3.79,
    },
    Object {
      "installment": 2,
      "mdr": 4.19,
    },
    Object {
      "installment": 7,
      "mdr": 4.59,
    },
  ],
  "transfer": 367,
}
`;

exports[`Account Reducer selectCompanyFees when company has other installments types should return the correct response 1`] = `
Object {
  "anticipation": 3.14,
  "antifraud": 70,
  "boleto": 380,
  "gateway": 50,
  "installments": Array [],
  "transfer": 367,
}
`;
